### PR TITLE
feat(lsp): cpu watchdog thread

### PR DESCRIPTION
**Pull request by nayeemrmn**

### CODE REVIEW

File: cli/Cargo.toml [hunk 1]
```diff
@@ -118,6 +118,7 @@ rustyline-derive = "=0.7.0"
 serde.workspace = true
 serde_repr.workspace = true
 shell-escape = "=0.1.5"
+sysinfo = "=0.29.10"
```
Comment (bartlomieju) [inline cli/Cargo.toml:121]: Discussed offline - this is adding another dynamic library on startup. Waiting on @littledivy's review

File: cli/lsp/language_server.rs [hunk 2]
```diff
@@ -3108,6 +3114,44 @@ impl tower_lsp::LanguageServer for LanguageServer {
...
+        loop {
+          thread::sleep(std::time::Duration::from_secs(10));
+          system.refresh_process_specifics(
+            pid,
+            ProcessRefreshKind::new().with_cpu(),
+          );
+          let Some(process) = system.process(pid) else {
+            continue;
+          };
+          if process.cpu_usage() > 90.0 {
```
[the remark this answers is not in the document]
Comment (bartlomieju) [reply cli/lsp/language_server.rs:3141]: Good point 👍 I think we should make it into a struct that gets a callback to measure the CPU usage and a callback to decide what the next interval should be - that way we can create a "test environment" that runs in a few seconds and check if the watchdog is behaving properly. I believe we have similar setup in the "update checker" that we can take inspiration from.

File: cli/Cargo.toml (hunk 1 above)
Comment (littledivy) [reply cli/Cargo.toml:121]: Looking at the usage I think we can manually implement APIs like we did in https://github.com/denoland/deno/pull/16441 

To keep it simple, use sysinfo crate as non-macOS dependency and have a Apple-specific implementation:

https://github.com/GuillaumeGomez/sysinfo/blob/84c3b36a730a91229dd5a738c725aac81a08c7c8/src/unix/apple/macos/process.rs#L663-L692

**Comment by nayeemrmn** [comment]
Got it green. It would be good if a reviewer on macos could check the cpu usage calculation by adding a `dbg!()` here and then running the extension under this patch:
```diff
diff --git a/cli/lsp/language_server.rs b/cli/lsp/language_server.rs
index 976b27c79..558ceb63c 100644
--- a/cli/lsp/language_server.rs
+++ b/cli/lsp/language_server.rs
@@ -208,6 +208,7 @@ impl CpuUsageWatchdog {
         const USAGE_THRESHOLD: usize = 90;
         const MAX_OVERUSE_PERIOD: usize = 20;
         let cpu_usage = self.state.refresh_cpu_usage();
+        dbg!(cpu_usage);
         if self.is_overusing {
           if cpu_usage > USAGE_THRESHOLD as f32 {
             self.overuse_interval_count += 1;
```

File: Releases.md
Comment (bartlomieju) [inline Releases.md:1]: This change looks unrelated

File: Cargo.lock
Comment (bartlomieju) [inline Cargo.lock:1]: Where are these changes coming from? Maybe the lockfile needs to be reset to the state of `main` branch?

File: runtime/ops/os/sys_info.rs
Comment (bartlomieju) [inline runtime/ops/os/sys_info.rs:1]: @mmastrac we could use you eyes in here - all of these APIs are added to avoid adding `sysinfo` dependency which adds more dynamic libs on macOS impacting startup time.

File: Cargo.lock
Comment (nayeemrmn) [reply Cargo.lock:1]: `runtime/Cargo.toml` adds sysinfo for non-macos

File: Releases.md
Comment (nayeemrmn) [reply Releases.md:1]: I just ran the formatter, apparently this file isn't checked or formatted on main

File: Cargo.lock [hunk 3]
```diff
@@ -4086,6 +4120,26 @@ dependencies = [
  "getrandom",
 ]
 
+[[package]]
+name = "rayon"
```
Comment (ry) [inline Cargo.lock:4124]: It’s unfortunate to add a dependency like this for such a small feature. Is this not avoidable somehow?

File: cli/lsp/language_server.rs [hunk 4]
```diff
@@ -176,6 +179,64 @@ pub struct StateSnapshot {
   pub npm: Option<StateNpmSnapshot>,
 }
 
+struct CpuUsageWatchdog {
```
Comment (ry) [inline cli/lsp/language_server.rs:182]: Maybe this should be in a separate file? Maybe we should cargo feature flag this ?

**Comment by nayeemrmn** [comment]
We're opting out of this in favour of direct perf improvements
